src/core/etl/src/Flow/ETL/PHP/Type/TypeFactory.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flow\ETL\PHP\Type;
+
+use Flow\ETL\Exception\InvalidArgumentException;
+use Flow\ETL\PHP\Type\Logical\List\ListElement;
+use Flow\ETL\PHP\Type\Logical\ListType;
+use Flow\ETL\PHP\Type\Logical\Map\MapKey;
+use Flow\ETL\PHP\Type\Logical\Map\MapValue;
+use Flow\ETL\PHP\Type\Logical\MapType;
+use Flow\ETL\PHP\Type\Logical\Structure\StructureElement;
+use Flow\ETL\PHP\Type\Logical\StructureType;
+use Flow\ETL\PHP\Type\Native\ArrayType;
+use Flow\ETL\PHP\Type\Native\NullType;
+use Flow\ETL\PHP\Type\Native\ObjectType;
+use Flow\ETL\PHP\Type\Native\ScalarType;
+
+final class TypeFactory
+{
+    public function getType(mixed $value) : Type
+    {
+        if (null === $value) {
+            return new NullType();
+        }
+
+        if (\is_scalar($value)) {
+            return ScalarType::fromString(\gettype($value));
+        }
+
+        if (\is_object($value)) {
+            return ObjectType::of($value::class);
+        }
+
+        if (\is_array($value)) {
+            if ([] === \array_filter($value, fn ($value) : bool => null !== $value)) {
```

Comment:
this logic should be covered by detector that should not return true on isMap/isList/isStruct, when array is empty or its an array of nulls.